u_pre_shared_key.go

```diff
@@ -0,0 +1,503 @@
+	b[1] = byte(extensionPreSharedKey)
+	b[2] = byte((e.Len() - 4) >> 8)
+	b[3] = byte(e.Len() - 4)
+
+	// identities length
+	identitiesLength := 0
+	for _, identity := range e.Identities {
+		identitiesLength += 2 + len(identity.Label) + 4 // identity length + identity + obfuscated ticket age
+	}
+	b[4] = byte(identitiesLength >> 8)
+	b[5] = byte(identitiesLength)
+
+	// identities
+	offset := 6
+	for _, identity := range e.Identities {
+		b[offset] = byte(len(identity.Label) >> 8)
+		b[offset+1] = byte(len(identity.Label))
+		offset += 2
+		copy(b[offset:], identity.Label)
+		offset += len(identity.Label)
+		b[offset] = byte(identity.ObfuscatedTicketAge >> 24)
+		b[offset+1] = byte(identity.ObfuscatedTicketAge >> 16)
+		b[offset+2] = byte(identity.ObfuscatedTicketAge >> 8)
+		b[offset+3] = byte(identity.ObfuscatedTicketAge)
+		offset += 4
+	}
+
+	cipherSuite := cipherSuiteTLS13ByID(e.CipherSuite)
+	if cipherSuite == nil {
+		return 0, errors.New("tls: cipher suite not supported")
+	}
+	earlySecret := cipherSuite.extract(e.SessionSecret, nil)
+	binderKey := cipherSuite.deriveSecret(earlySecret, resumptionBinderLabel, nil)
+
+	// concatenate ClientHello and PreSharedKeyExtension
+	rawHelloSoFar := append(raw, b[:offset]...)
+	transcript := cipherSuite.hash.New()
+	transcript.Write(rawHelloSoFar)
+	e.binders = [][]byte{cipherSuite.finishedHash(binderKey, transcript)}
+
+	// binders length
+	bindersLength := 0
+	for _, binder := range e.binders {
+		bindersLength += len(binder) + 1 // binder length + binder
+	}
+	b[offset] = byte(bindersLength >> 8)
+	b[offset+1] = byte(bindersLength)
+	offset += 2
+
+	// binders
+	for _, binder := range e.binders {
+		b[offset] = byte(len(binder))
+		offset++
+		copy(b[offset:], binder)
+		offset += len(binder)
+	}
+
+	return e.Len(), io.EOF
+}
+
+func (e *FakePreSharedKeyExtension) Binders() [][]byte {
+	return nil
+}
+
+func (e *FakePreSharedKeyExtension) UnmarshalJSON(data []byte) error {
+	var pskAccepter struct {
+		CipherSuite   uint16        `json:"cipher_suite"`
+		SessionSecret []byte        `json:"session_secret"`
+		Identities    []PskIdentity `json:"identities"`
+	}
+
+	if err := json.Unmarshal(data, &pskAccepter); err != nil {
+		return err
+	}
+
+	e.CipherSuite = pskAccepter.CipherSuite
+	e.SessionSecret = pskAccepter.SessionSecret
+	e.Identities = pskAccepter.Identities
+	return nil
+}
+
+// HardcodedPreSharedKeyExtension is an extension used to set the PSK extension in the
+// ClientHello.
+//
+// It does not compute binders based on ClientHello, but uses the binders specified instead.
+//
+// TODO: Only one of FakePreSharedKeyExtension and HardcodedPreSharedKeyExtension should
+// be kept, the other one should be just removed. We still need to learn more of the safety
+// of hardcoding both Identities and Binders without recalculating the latter.
+type HardcodedPreSharedKeyExtension struct {
+	Identities []PskIdentity `json:"identities"`
+	Binders    [][]byte      `json:"binders"`
+}
```

Comment:
`HardcodePreSharedKeyExtension` should embed `UnimplementedPreSharedKeyExtension`. I will push a change. 
```suggestion
type HardcodedPreSharedKeyExtension struct {
        UnimplementedPreSharedKeyExtension

	Identities []PskIdentity `json:"identities"`
	Binders    [][]byte      `json:"binders"`
}
```